### DISCUSSION TITLE
isHandTracking remaining false in AR session

### DISCUSSION THREAD

**Discussion by rekliner** [pmndrs/xr#207] [Q&A] [answered]
I wanted to see if anyone had any suggestions for this before I formally submitted an issue.  Basically it seems that react-xr is failing to set up hand tracking in AR sessions.  Anyone else struggled with this?

The isHandTracking state in useXR is remaining false for the AR session using hand-tracking whereas flipping over to a VR session works as it should.  The XRInputSource property "hand" is still null in AR while it properly contains an XRHand object in VR.    When using the same AR hand-tracking in raw three.js the XRHand object is present in the session.InputSources object as it should be.  

I'm using the latest Oculus browser to test this.
See https://e4cx0r.csb.app/  for a react-xr example of failed AR hand-tracking.  useEffect shows the session from useXR in the console.  Note that XRSession.InputSources[0].hand === null.  The only difference I can see in the controllers is that XRController.InputSource.gamepad is null for hands but contains the buttons/joys for a controller.
See https://lv25uw.csb.app/ for a three.js example properly registering hand-tracking in AR.  The session is logged in the console here too.  Note that XRSession.InputSources[0].hand === XRHand.

To get as deep as I can with it: state is being set here: https://github.com/pmndrs/react-xr/blob/fdd3e54794d0171d75d9346c14a6b82a754353f0/src/XR.tsx#L134 

perhaps cleanup is being triggered early somehow and removing the listener for "inputsourceschange":
https://github.com/pmndrs/react-xr/blob/fdd3e54794d0171d75d9346c14a6b82a754353f0/src/XR.tsx#L150  
that's a close as I can get to a suggestion for a PR.

**Comment by CodyJasonBennett** [comment]
I've found that we can sometimes be too late to session init and three goes out of sync, where I eagerly call sessionstart handlers: https://github.com/pmndrs/react-xr/blob/43bdb9fba56d74b8176024618694f26b2aa3cf03/src/XR.tsx#L143-L144

I think I'd need to do the same for internal input handlers, but fixing this desync problem at its core would be ideal.

**Reply by rekliner** [reply]
So adding an eager check along with the session start would be the band-aid solution?
`if (Object.values(session?.inputSources).some((source) => source.hand)) handleInputSourcesChange({ type: 'inputsourceschange', target: session })`  

PR #208 is created

**Comment by CodyJasonBennett** [comment, marked as answer]
Fixed in https://github.com/pmndrs/react-xr/pull/208.
